### ISSUE TITLE
Getting "container_create_or_lookup_app_group_path_by_app_group_identifier: client is not entitled" when running React Native app from XCode

### ISSUE THREAD

**Issue by irina-unique** [msintuneappsdk/ms-intune-app-sdk-ios#398]
**Describe the issue:**
Getting `container_create_or_lookup_app_group_path_by_app_group_identifier: client is not entitled` when running React Native app from XCode.
It looks like `registerAndEnrollAccount` finishes successfully (not errors are thrown) but trying to get the `enrolledAccount` returns null.

**To Reproduce
Steps to reproduce the behavior:**
1. Follow the instructions provided in the official documentation for setting up the iOS app
2. Run the app from XCode
3. The following logs are shown

**Screenshots and logs:**
![image](https://github.com/msintuneappsdk/ms-intune-app-sdk-ios/assets/137764243/3592ab3b-65f4-4433-ab1a-68ea88bf9e63)

**Smartphone**:
Device: Simulator - iPhone 15 Pro
OS: Latest
Intune App SDK for iOS (please complete the following information):

What version of the Intune SDK are you using? Are you using the latest version?
Latest Version

**Further info after running** `react-native-info`:
`System:
  OS: macOS 14.2.1
  CPU: (12) arm64 Apple M2 Max
  Memory: 90.30 MB / 32.00 GB
  Shell:
    version: "5.9"
    path: /bin/zsh
Binaries:
  Node:
    version: 21.5.0
    path: /opt/homebrew/bin/node
  Yarn: Not Found
  npm:
    version: 10.2.4
    path: /opt/homebrew/bin/npm
  Watchman:
    version: 2023.12.04.00
    path: /opt/homebrew/bin/watchman
Managers:
  CocoaPods:
    version: 1.14.3
    path: /opt/homebrew/bin/pod
SDKs:
  iOS SDK:
    Platforms:
      - DriverKit 23.2
      - iOS 17.2
      - macOS 14.2
      - tvOS 17.2
      - watchOS 10.2
  Android SDK: Not Found
IDEs:
  Android Studio: 2023.1 AI-231.9392.1.2311.11076708
  Xcode:
    version: 15.1/15C65
    path: /usr/bin/xcodebuild
Languages:
  Java:
    version: 20.0.2
    path: /usr/bin/javac
  Ruby:
    version: 3.2.2
    path: /opt/homebrew/opt/ruby/bin/ruby
npmPackages:
  "@react-native-community/cli": Not Found
  react: Not Found
  react-native: Not Found
  react-native-macos: Not Found
npmGlobalPackages:
  "*react-native*": Not Found
Android:
  hermesEnabled: true
  newArchEnabled: false
iOS:
  hermesEnabled: true
  newArchEnabled: false`

**Comment by darosal** [comment]
Hi @irina-unique, thank you for reaching out. 

The `container_create_or_lookup_app_group_path_by_app_group_identifier: client is not entitled` is an expected log statement and isn't indicative of any issues.

The registerAndEnrollAccount API kicks off an asynchronous bit of code that attempts to enroll the account. To see the outcome of the enrollment attempt, please implement the IntuneMAMEnrollmentDelegate. Here are our public docs with more information on how to do so: https://learn.microsoft.com/mem/intune/developer/app-sdk-ios-phase3#status-result-and-debug-notifications.

Are you seeing an errorString as part of the status returned from the `enrollmentRequestWithStatus:` delegate method?

**Comment by irina-unique** [comment]
hi @darosal, thank you for the quick response! Unfortunately I did not get any progress.
**What I tried:**
- implementing the debugging methods in the native module and listen to them in JavaScript
- added logs directly in the module (this in case the communication with JavaScript doesn't work - so it is a back up)
- building the app in different ways (profiles certs), to make sure that it is not an issue with these
- debugged on the simulator, debugged on device (iPhone 13 Pro), created a TestFlight build and also tried from there
- debugged both development and release builds

**Results:**
- no `errorString`/`identity`/`statusCode` displayed
- no `enrolledAccount` (I understood that this is async and not necessarily an issue, but good to mention)
- I see nothing in Intune Admin Portal, and this leads to ->

**Questions:**
- What should I actually see in Intune Admin Portal?
- Should I see anything in the mobile app? (I expect yes, but I don't see any pop-ups. I only have the browser opening for the MSAL login)

**Comment by darosal** [comment]
Hi @irina-unique, is the `enrollmentRequestWithStatus:` delegate method being hit at all? Or is it being called without an error?

Just as a sanity check, has the delegate implementation been set to the IntuneMAMEnrollmentManager's delegate property?

**Comment by irina-unique** [comment]
hi @darosal, good check! I will add here the code for the native modules:

**IntuneModule.h**
```
//
//  IntuneModule.h
//  UniqueMoments
//
//
#import <React/RCTBridgeModule.h>
#import <React/RCTEventEmitter.h>
#import <IntuneMAMSwift/IntuneMAMEnrollmentDelegate.h>

#ifndef IntuneModule_h
#define IntuneModule_h

#endif /* IntuneModule_h */

@interface IntuneModule : RCTEventEmitter <RCTBridgeModule, IntuneMAMEnrollmentDelegate>

@end
```

**IntuneModule.m**
```
//
//  IntuneModule.m
//  UniqueMoments
//
//

#import <Foundation/Foundation.h>
#import "IntuneModule.h"
#import <IntuneMAMSwift/IntuneMAMEnrollmentManager.h>
#import <IntuneMAMSwift/IntuneMAMPolicyManager.h>
#import <IntuneMAMSwift/IntuneMAMPolicy.h>
#import <React/RCTEventEmitter.h>

@implementation IntuneModule

RCT_EXPORT_MODULE();

+ (void)initialize {
    // This is called exactly once, before the first method of this class is called.
    [IntuneMAMEnrollmentManager initialize];
}

- (instancetype)init {
  self = [super init];
  if (self) {
    // Set this instance as the delegate for the IntuneMAMEnrollmentManager
    [IntuneMAMEnrollmentManager instance].delegate = self;
  }
  return self;
}

RCT_EXPORT_METHOD(registerAndEnrollAccount:(NSString *)identity callback:(RCTResponseSenderBlock)callback) {
    NSLog(@"CUSTOM INTUNE MODULE - registerAndEnrollAccount with params: %@", identity);

    // Ensure identity is not nil or empty
    if (!identity || [identity isEqualToString:@""]) {
        NSString *errorMessage = @"Invalid identity";
        callback(@[errorMessage]);
        return;
    }

    // Attempt to register and enroll
    [IntuneMAMEnrollmentManager.instance registerAndEnrollAccount:identity];
    NSLog(@"CUSTOM INTUNE MODULE - registerAndEnrollAccount with params2: %@", identity);

    // Assume enrollment successful if no error occurred during registration
    callback(@[[NSNull null]]);
}

RCT_EXPORT_METHOD(getRegisteredAccounts) {
    // Call the corresponding method to get registered accounts
    // NSArray *registeredAccounts = [IntuneMAMEnrollmentManager.instance registeredAccounts];
    // callback(@[registeredAccounts]);
    NSString *registeredUser = [IntuneMAMEnrollmentManager instance].enrolledAccount;
    NSLog(@"CUSTOM INTUNE MODULE - registeredUser: %@", registeredUser);
}


RCT_EXPORT_METHOD(getRegisteredAccountIds:(RCTResponseSenderBlock)callback) {
    // Call the corresponding method to get registered account IDs
    NSArray *registeredAccountIds = [IntuneMAMEnrollmentManager.instance registeredAccountIds];
    callback(@[registeredAccountIds]);
}

RCT_EXPORT_METHOD(getEnrolledAccount:(RCTResponseSenderBlock)callback) {
    // Call the corresponding method to get enrolled account
    NSString *enrolledAccount = [IntuneMAMEnrollmentManager.instance enrolledAccount];
    callback(@[enrolledAccount ?: [NSNull null]]);
}

- (void)enrollmentRequestWithStatus:(IntuneMAMEnrollmentStatus *)status
{
    NSLog(@"<DEBUG> CUSTOM INTUNE MODULE - enrollment result for identity %@ with status code %ld", status.identity, (unsigned long)status.statusCode);
    NSLog(@"<DEBUG> CUSTOM INTUNE MODULE - Debug Message: %@", status.errorString);
    [self sendEventWithName:@"EnrollmentStatus" body:@{@"status": status.errorString}];
}

- (void)policyRequestWithStatus:(IntuneMAMEnrollmentStatus *)status
{
    NSLog(@"<DEBUG> CUSTOM INTUNE MODULE - policy check-in result for identity %@ with status code %ld", status.identity, (unsigned long)status.statusCode);
    NSLog(@"<DEBUG> CUSTOM INTUNE MODULE - Debug Message: %@", status.errorString);
    [self sendEventWithName:@"PolicyStatus" body:@{@"status": status.errorString}];
}

- (void)unenrollRequestWithStatus:(IntuneMAMEnrollmentStatus*)status
{
    NSLog(@"<DEBUG> CUSTOM INTUNE MODULE - un-enroll result for identity %@ with status code %ld", status.identity, (unsigned long)status.statusCode);
    NSLog(@"<DEBUG> CUSTOM INTUNE MODULE - Debug Message: %@", status.errorString);
    [self sendEventWithName:@"UnenrollmentStatus" body:@{@"status": status.errorString}];
}

- (NSArray<NSString *> *)supportedEvents {
    return @[@"EnrollmentStatus", @"PolicyStatus", @"UnenrollmentStatus"];
}

@end
```

**Comment by darosal** [comment]
Hi @irina-unique, thanks for providing that info. Can Intune logs be collected and emailed to me at darosal@microsoft.com?

To collect them, the IntuneMAMDiagnosticConsole needs to be called from within the app. Maybe a test button to trigger this flow can be added temporarily just to facilitate this log collection. Here are the docs around troubleshooting: https://learn.microsoft.com/mem/intune/developer/app-sdk-ios-appendix#how-can-i-troubleshoot-my-app

And just to clean out any leftover state from the various tests, can the simulator be wiped before reproing and collecting the logs?

**Comment by irina-unique** [comment]
Hi @darosal, thank you for providing further support! I have sent an email with the logs. Also, while implementing this, I finally got some logs from the EnrollmentStatus. I now continue looking into this:

![image](https://github.com/msintuneappsdk/ms-intune-app-sdk-ios/assets/137764243/28946a39-4a58-48aa-9785-8145a5d00e0b)

**Comment by darosal** [comment]
Great thank you for sending that along! The 217 error code does shed some more light on what could be happening. 

Can `displayDiagnosticConsole` be called on the IntuneMAMDiagnosticConsole? There's more detailed logs that can be gathered from the presented UI that will have more information about why the AAD token is not setup properly. Could the log file from that diagnostic console be emailed to me as well?

Given that something is wrong with the AAD token, can you double check that all the steps in [this doc](https://learn.microsoft.com/mem/intune/developer/app-sdk-ios-phase2) have been followed?

**Comment by irina-unique** [comment]
@darosal Thank you very much for checking the logs and providing further support!

I doubled-check [this doc](https://learn.microsoft.com/mem/intune/developer/app-sdk-ios-phase2) again and I couldn't find anything that we missed. The app configuration is the same as on the app registration, the necessary permissions are granted, and the MS login works generally as we successfully get back to the app with a valid token. We are using [this library](https://www.npmjs.com/package/react-native-msal) for implementing MS login. From what I checked it is what we need - we chose to use it as it saves us from building another native module (and avoiding other implementation errors). Do you think this can be an issue?

The `displayDiagnosticConsole` is called on the IntuneMAMDiagnosticConsole, but I see no diagnostic console opening. I tried to debug, but I am not sure what the issue is. I tried the following:
```
RCT_EXPORT_METHOD(displayDiagnosticConsole)
{
    dispatch_async(dispatch_get_main_queue(), ^{
        @try {
            [IntuneMAMDiagnosticConsole displayDiagnosticConsole];
        }
        @catch (NSException *exception) {
            NSLog(@"Error displaying Intune diagnostic console: %@", exception);
        }
    });
}
```
and simply
```
RCT_EXPORT_METHOD(displayDiagnosticConsole)
{
    NSLog(@"<CONOSLE-DEBUG> CUSTOM INTUNE MODULE - displayDiagnosticConsole");
    [IntuneMAMDiagnosticConsole displayDiagnosticConsole];
}
```

**Comment by glenwang88** [comment]
To: @irina-unique  CC: @darosal 
Have you got the MAM policy applied to your app? If still failed, as the Diagnostic Console did not work for you, you can use below steps to collect MAM logs:

1) Backup the device to the local computer using iTunes or Finder
2) Launch iExplorer (https://www.macroplant.com/iexplorer/)
3) Select iExplorer->Preferences menu and check ‘Show hidden files and folders’
4) Find the last backup under ‘Browse iTune Backups’
5) Under ‘Backup Explorer’ find App->{AIP_BUNDLE_ID} 
6) Export all files under Library/.IntuneMAM

If you can collect the logs, or if you have issue with the procedure, then could you let me know your email I will connect.

**Comment by irina-unique** [comment]
hey @glenwang88, thank you for coming back to me!

In the meantime we managed to get the app to work.
I am not sure exactly which change was the one that got the enrolment working. We noticed that the `openURL` callback was not working and we changed a bit the native code, so it might be this.

Also, what was really helpful, was to play around with this app: https://github.com/msintuneappsdk/Chatr-Sample-Intune-iOS-App/.

I will close the issue as it is working now. Thank you very much for the support @glenwang88 and @darosal!
